File: ios/Jamboree/Games/SecretHitler/SecretHitlerEngine.swift
import Foundation

enum SecretHitlerPhase: String {
    case lobby, nomination, election
    case presidentDiscard, chancellorEnact, vetoDecision
    case policyPeek, investigation, investigationReveal
    case specialElection, execution
    case gameOver
}

enum SecretHitlerRole: String { case liberal, fascist, hitler }
enum SecretHitlerParty: String { case liberal, fascist }
enum SecretHitlerPolicy: String { case liberal, fascist }
enum SecretHitlerWinner: String { case liberal, fascist }
enum SecretHitlerWinReason: String {
    case fiveLiberalPolicies, sixFascistPolicies, hitlerElectedChancellor, hitlerExecuted
}

extension SecretHitlerRole {
    var party: SecretHitlerParty { self == .liberal ? .liberal : .fascist }
}

final class SecretHitlerPlayer {
    let id: String
    let name: String
    let isHost: Bool
    var role: SecretHitlerRole?
    var alive: Bool = true
    init(id: String, name: String, isHost: Bool) {
        self.id = id; self.name = name; self.isHost = isHost
    }
}

/// Pure game logic for Secret Hitler — supports 5–10 players.
/// Network/UI lives in the server adapter and view; the engine never touches I/O.
final class SecretHitlerEngine {
    private var rng: any RandomNumberGenerator

    private(set) var players: [String: SecretHitlerPlayer] = [:]
    /// Stable seating order — president rotates along this list.
    private(set) var seatOrder: [String] = []

    var phase: SecretHitlerPhase = .lobby

    // Round state
    private(set) var presidentId: String?
    private(set) var chancellorNomineeId: String?
    private(set) var chancellorId: String?
    private(set) var previousPresidentId: String?
    private(set) var previousChancellorId: String?
    /// When a special-election president finishes their term, the rotation
    /// returns to the seat after this id.
    private(set) var specialElectionResumeAfter: String?

    private(set) var electionTracker: Int = 0
    private(set) var liberalPolicies: Int = 0
    private(set) var fascistPolicies: Int = 0
    private(set) var vetoUnlocked = false

    // Decks
    private(set) var drawPile: [SecretHitlerPolicy] = []
    private(set) var discardPile: [SecretHitlerPolicy] = []
    private(set) var presidentialHand: [SecretHitlerPolicy] = []
    private(set) var chancellorHand: [SecretHitlerPolicy] = []
    private(set) var vetoRequested = false

    // Voting / power state
    var electionVotes: [String: Bool] = [:]
    private(set) var lastElectionPassed: Bool?
    private(set) var lastEnactedPolicy: SecretHitlerPolicy?
    private(set) var lastEnactedByChaos = false
    private(set) var lastExecutedId: String?
    private(set) var peekedPolicies: [SecretHitlerPolicy] = []
    private(set) var pendingInvestigationId: String?
    private(set) var lastInvestigation: (subjectId: String, party: SecretHitlerParty)?
    private(set) var investigatedIds: Set<String> = []

    var winner: SecretHitlerWinner?
    var winReason: SecretHitlerWinReason?

    let tutorialVote = TutorialVote()

    init(rng: any RandomNumberGenerator = SystemRandomNumberGenerator()) { self.rng = rng }

    // MARK: Lobby
    @discardableResult
    func addPlayer(id: String, name: String, isHost: Bool = false) -> SecretHitlerPlayer {
        let p = SecretHitlerPlayer(id: id, name: name, isHost: isHost)
        players[id] = p
        if !seatOrder.contains(id) { seatOrder.append(id) }
        return p
    }

    func removePlayer(_ id: String) {
        guard phase == .lobby else { return }
        players[id] = nil
        seatOrder.removeAll { $0 == id }
    }

    var canStart: Bool { phase == .lobby && (5...10).contains(players.count) }

    var alive: [SecretHitlerPlayer] {
        seatOrder.compactMap { players[$0] }.filter { $0.alive }
    }

    func start() {
        guard canStart else { return }
        assignRoles()
        buildDeck()
        seatOrder = seatOrder.shuffled(using: &rng)
        presidentId = seatOrder.first
        phase = .nomination
    }

    private func assignRoles() {
        let n = seatOrder.count
        let liberalCount: Int = [5: 3, 6: 4, 7: 4, 8: 5, 9: 5, 10: 6][n] ?? 3
        let fascistCount = n - liberalCount - 1 // excludes Hitler
        var roles: [SecretHitlerRole] = Array(repeating: .liberal, count: liberalCount)
        roles += Array(repeating: .fascist, count: fascistCount)
        roles.append(.hitler)
        roles.shuffle(using: &rng)
        for (i, pid) in seatOrder.enumerated() { players[pid]?.role = roles[i] }
    }

    private func buildDeck() {
        var deck: [SecretHitlerPolicy] = Array(repeating: .liberal, count: 6)
                                       + Array(repeating: .fascist, count: 11)
        deck.shuffle(using: &rng)
        drawPile = deck
        discardPile = []
    }

    /// Fascists know each other & Hitler. Hitler additionally knows the
    /// fascists only in 5–6 player games (canonical rule).
    func knownAllies(for playerId: String) -> [String] {
        guard let me = players[playerId], let role = me.role else { return [] }
        let n = seatOrder.count
        switch role {
        case .liberal: return []
        case .fascist:
            return seatOrder.filter { id in
                guard id != playerId, let p = players[id], let r = p.role else { return false }
                return r == .fascist || r == .hitler
            }
        case .hitler:
            if n >= 7 { return [] }
            return seatOrder.filter { id in
                guard id != playerId, let p = players[id] else { return false }
                return p.role == .fascist
            }
        }
    }

    // MARK: Nomination
    func eligibleChancellorNominees() -> [SecretHitlerPlayer] {
        let aliveCount = alive.count
        return alive.filter { p in
            guard p.id != presidentId else { return false }
            if p.id == previousChancellorId { return false }
            if aliveCount > 5 && p.id == previousPresidentId { return false }
            return true
        }
    }

    @discardableResult
    func nominateChancellor(_ targetId: String) -> Bool {
        guard phase == .nomination,
              eligibleChancellorNominees().contains(where: { $0.id == targetId }) else { return false }
        chancellorNomineeId = targetId
        electionVotes.removeAll()
        phase = .election
        return true
    }

    // MARK: Election
    /// Returns true once every alive player has voted.
    @discardableResult
    func submitVote(voterId: String, ja: Bool) -> Bool {
        guard phase == .election,
              let v = players[voterId], v.alive else { return false }
        electionVotes[voterId] = ja
        return electionVotes.count >= alive.count
    }

    /// Resolves a finished vote. Returns whether the election passed.
    @discardableResult
    func resolveElection() -> Bool {
        guard phase == .election else { return false }
        let yes = electionVotes.values.filter { $0 }.count
        let no = electionVotes.values.filter { !$0 }.count
        let passed = yes > no
        lastElectionPassed = passed

        if passed {
            previousPresidentId = presidentId
            previousChancellorId = chancellorNomineeId
            chancellorId = chancellorNomineeId
            electionTracker = 0

            // Hitler-elected-chancellor win.
            if fascistPolicies >= 3,
               let cid = chancellorId, players[cid]?.role == .hitler {
                winner = .fascist
                winReason = .hitlerElectedChancellor
                phase = .gameOver
                return true
            }
            dealPresidentialHand()
            phase = .presidentDiscard
            return true
        } else {
            chancellorNomineeId = nil
            chancellorId = nil
            electionTracker += 1
            if electionTracker >= 3 {
                triggerChaos()
            } else {
                advancePresident()
                phase = .nomination
            }
            return false
        }
    }

    // MARK: Legislative
    private func dealPresidentialHand() {
        ensureDeckHasAtLeast(3)
        presidentialHand = Array(drawPile.prefix(3))
        drawPile.removeFirst(3)
    }

    @discardableResult
    func presidentDiscard(index: Int) -> Bool {
        guard phase == .presidentDiscard,
              presidentialHand.indices.contains(index) else { return false }
        var hand = presidentialHand
        discardPile.append(hand.remove(at: index))
        chancellorHand = hand
        presidentialHand = []
        vetoRequested = false
        phase = .chancellorEnact
        return true
    }

    @discardableResult
    func chancellorEnact(index: Int) -> Bool {
        guard phase == .chancellorEnact,
              chancellorHand.indices.contains(index) else { return false }
        var hand = chancellorHand
        let played = hand.remove(at: index)
        discardPile.append(contentsOf: hand)
        chancellorHand = []
        enact(policy: played, byChaos: false)
        return true
    }

    @discardableResult
    func chancellorRequestVeto() -> Bool {
        guard phase == .chancellorEnact, vetoUnlocked else { return false }
        vetoRequested = true
        phase = .vetoDecision
        return true
    }

    @discardableResult
    func presidentVetoResponse(confirm: Bool) -> Bool {
        guard phase == .vetoDecision else { return false }
        if confirm {
            discardPile.append(contentsOf: chancellorHand)
            chancellorHand = []
            vetoRequested = false
            // Veto counts as a failed government — advance the tracker.
            electionTracker += 1
            if electionTracker >= 3 {
                triggerChaos()
            } else {
                advancePresident()
                phase = .nomination
            }
        } else {
            vetoRequested = false
            phase = .chancellorEnact
        }
        return true
    }

    private func enact(policy: SecretHitlerPolicy, byChaos: Bool) {
        lastEnactedPolicy = policy
        lastEnactedByChaos = byChaos
        if policy == .liberal { liberalPolicies += 1 } else { fascistPolicies += 1 }
        if fascistPolicies >= 5 { vetoUnlocked = true }

        // Win checks first.
        if liberalPolicies >= 5 {
            winner = .liberal; winReason = .fiveLiberalPolicies; phase = .gameOver; return
        }
        if fascistPolicies >= 6 {
            winner = .fascist; winReason = .sixFascistPolicies; phase = .gameOver; return
        }

        if !byChaos, policy == .fascist, let power = presidentialPower() {
            enterPower(power)
        } else {
            advancePresident()
            phase = .nomination
        }
    }

    /// Canonical Secret Hitler board powers, by player count + fascist track position.
    enum Power { case peek, investigate, specialElection, execution }
    private func presidentialPower() -> Power? {
        switch seatOrder.count {
        case 5, 6:
            switch fascistPolicies {
            case 3: return .peek
            case 4, 5: return .execution
            default: return nil
            }
        case 7, 8:
            switch fascistPolicies {
            case 2: return .investigate
            case 3: return .specialElection
            case 4, 5: return .execution
            default: return nil
            }
        case 9, 10:
            switch fascistPolicies {
            case 1, 2: return .investigate
            case 3: return .specialElection
            case 4, 5: return .execution
            default: return nil
            }
        default: return nil
        }
    }

    private func enterPower(_ p: Power) {
        switch p {
        case .peek:
            ensureDeckHasAtLeast(3)
            peekedPolicies = Array(drawPile.prefix(3))
            phase = .policyPeek
        case .investigate:
            phase = .investigation
        case .specialElection:
            phase = .specialElection
        case .execution:
            phase = .execution
        }
    }

    // MARK: Powers
    @discardableResult
    func acknowledgePeek() -> Bool {
        guard phase == .policyPeek else { return false }
        peekedPolicies = []
        advancePresident()
        phase = .nomination
        return true
    }

    /// President investigates a target — they may not investigate themselves
    /// or anyone already investigated this game.
    func investigationTargets() -> [SecretHitlerPlayer] {
        alive.filter { $0.id != presidentId && !investigatedIds.contains($0.id) }
    }

    @discardableResult
    func investigate(targetId: String) -> Bool {
        guard phase == .investigation,
              let t = players[targetId], t.alive,
              t.id != presidentId, !investigatedIds.contains(targetId) else { return false }
        pendingInvestigationId = targetId
        lastInvestigation = (targetId, t.role?.party ?? .liberal)
        investigatedIds.insert(targetId)
        phase = .investigationReveal
        return true
    }

    @discardableResult
    func acknowledgeInvestigation() -> Bool {
        guard phase == .investigationReveal else { return false }
        pendingInvestigationId = nil
        advancePresident()
        phase = .nomination
        return true
    }

    /// Special election: the named player becomes the next president for one
    /// round, then the rotation returns to the seat after the original.
    @discardableResult
    func callSpecialElection(targetId: String) -> Bool {
        guard phase == .specialElection,
              let t = players[targetId], t.alive, t.id != presidentId else { return false }
        specialElectionResumeAfter = presidentId
        presidentId = targetId
        chancellorNomineeId = nil
        chancellorId = nil
        phase = .nomination
        return true
    }

    /// President executes a player. Killing Hitler ends the game (Liberals win).
    func executionTargets() -> [SecretHitlerPlayer] {
        alive.filter { $0.id != presidentId }
    }

    @discardableResult
    func executePlayer(targetId: String) -> Bool {
        guard phase == .execution,
              let t = players[targetId], t.alive, t.id != presidentId else { return false }
        t.alive = false
        lastExecutedId = targetId
        if t.role == .hitler {
            winner = .liberal; winReason = .hitlerExecuted; phase = .gameOver
            return true
        }
        // If the executed player was the previous chancellor/president, those
        // term limits no longer constrain — the canonical rule treats dead
        // players as not term-limiting.
        if previousChancellorId == targetId { previousChancellorId = nil }
        if previousPresidentId == targetId { previousPresidentId = nil }
        advancePresident()
        phase = .nomination
        return true
    }

    // MARK: Rotation / chaos
    private func advancePresident() {
        if let resume = specialElectionResumeAfter {
            specialElectionResumeAfter = nil
            presidentId = nextAlive(after: resume)
        } else if let p = presidentId {
            presidentId = nextAlive(after: p)
        }
        chancellorNomineeId = nil
        chancellorId = nil
    }

    private func nextAlive(after id: String) -> String? {
        guard let idx = seatOrder.firstIndex(of: id) else { return alive.first?.id }
        let n = seatOrder.count
        for offset in 1...n {
            let candidate = seatOrder[(idx + offset) % n]
            if let p = players[candidate], p.alive { return candidate }
        }
        return nil
    }

    private func triggerChaos() {
        ensureDeckHasAtLeast(1)
        let top = drawPile.removeFirst()
        enact(policy: top, byChaos: true)
        if phase == .gameOver { return }
        electionTracker = 0
        previousChancellorId = nil
        previousPresidentId = nil
        // `enact` already advanced the president because powers are skipped on chaos.
    }

    private func ensureDeckHasAtLeast(_ k: Int) {
        if drawPile.count >= k { return }
        var combined = drawPile + discardPile
        combined.shuffle(using: &rng)
        drawPile = combined
        discardPile = []
    }
}

File: android/app/src/main/java/com/example/jamboree/games/secrethitler/SecretHitlerEngine.kt
package com.example.ubapp.games.secrethitler

import kotlin.random.Random

enum class SecretHitlerPhase(val wire: String) {
    LOBBY("lobby"),
    NOMINATION("nomination"),
    ELECTION("election"),
    PRESIDENT_DISCARD("presidentDiscard"),
    CHANCELLOR_ENACT("chancellorEnact"),
    VETO_DECISION("vetoDecision"),
    POLICY_PEEK("policyPeek"),
    INVESTIGATION("investigation"),
    INVESTIGATION_REVEAL("investigationReveal"),
    SPECIAL_ELECTION("specialElection"),
    EXECUTION("execution"),
    GAME_OVER("gameOver"),
}

enum class SecretHitlerRole(val wire: String) {
    LIBERAL("liberal"), FASCIST("fascist"), HITLER("hitler");

    val party: SecretHitlerParty
        get() = if (this == LIBERAL) SecretHitlerParty.LIBERAL else SecretHitlerParty.FASCIST
}

enum class SecretHitlerParty(val wire: String) { LIBERAL("liberal"), FASCIST("fascist") }
enum class SecretHitlerPolicy(val wire: String) { LIBERAL("liberal"), FASCIST("fascist") }
enum class SecretHitlerWinner(val wire: String) { LIBERAL("liberal"), FASCIST("fascist") }
enum class SecretHitlerWinReason(val wire: String) {
    FIVE_LIBERAL_POLICIES("fiveLiberalPolicies"),
    SIX_FASCIST_POLICIES("sixFascistPolicies"),
    HITLER_ELECTED_CHANCELLOR("hitlerElectedChancellor"),
    HITLER_EXECUTED("hitlerExecuted"),
}

class SecretHitlerPlayer(val id: String, val name: String, val isHost: Boolean) {
    var role: SecretHitlerRole? = null
    var alive: Boolean = true
}

data class SecretHitlerInvestigation(val subjectId: String, val party: SecretHitlerParty)

/**
 * Pure game logic for Secret Hitler. Mirrors SecretHitlerEngine.swift —
 * supports 5–10 players. The server adapter feeds in events; the engine
 * never touches network or UI code.
 */
class SecretHitlerEngine(private val rng: Random = Random.Default) {
    val tutorialVote = com.example.ubapp.tutorials.TutorialVote()

    val players: MutableMap<String, SecretHitlerPlayer> = linkedMapOf()
    val seatOrder: MutableList<String> = mutableListOf()

    var phase: SecretHitlerPhase = SecretHitlerPhase.LOBBY

    var presidentId: String? = null; private set
    var chancellorNomineeId: String? = null; private set
    var chancellorId: String? = null; private set
    var previousPresidentId: String? = null; private set
    var previousChancellorId: String? = null; private set
    var specialElectionResumeAfter: String? = null; private set

    var electionTracker: Int = 0; private set
    var liberalPolicies: Int = 0; private set
    var fascistPolicies: Int = 0; private set
    var vetoUnlocked: Boolean = false; private set
    var vetoRequested: Boolean = false; private set

    val drawPile: MutableList<SecretHitlerPolicy> = mutableListOf()
    val discardPile: MutableList<SecretHitlerPolicy> = mutableListOf()
    var presidentialHand: List<SecretHitlerPolicy> = emptyList(); private set
    var chancellorHand: List<SecretHitlerPolicy> = emptyList(); private set

    val electionVotes: MutableMap<String, Boolean> = mutableMapOf()
    var lastElectionPassed: Boolean? = null; private set
    var lastEnactedPolicy: SecretHitlerPolicy? = null; private set
    var lastEnactedByChaos: Boolean = false; private set
    var lastExecutedId: String? = null; private set
    var peekedPolicies: List<SecretHitlerPolicy> = emptyList(); private set
    var pendingInvestigationId: String? = null; private set
    var lastInvestigation: SecretHitlerInvestigation? = null; private set
    val investigatedIds: MutableSet<String> = mutableSetOf()

    var winner: SecretHitlerWinner? = null
    var winReason: SecretHitlerWinReason? = null

    fun addPlayer(id: String, name: String, isHost: Boolean = false): SecretHitlerPlayer {
        val p = SecretHitlerPlayer(id, name, isHost)
        players[id] = p
        if (id !in seatOrder) seatOrder.add(id)
        return p
    }

    fun removePlayer(id: String) {
        if (phase != SecretHitlerPhase.LOBBY) return
        players.remove(id); seatOrder.remove(id)
    }

    val canStart: Boolean
        get() = phase == SecretHitlerPhase.LOBBY && players.size in 5..10

    val alive: List<SecretHitlerPlayer>
        get() = seatOrder.mapNotNull { players[it] }.filter { it.alive }

    fun start() {
        if (!canStart) return
        assignRoles()
        buildDeck()
        seatOrder.shuffle(rng)
        presidentId = seatOrder.firstOrNull()
        phase = SecretHitlerPhase.NOMINATION
    }

    private fun assignRoles() {
        val n = seatOrder.size
        val liberalCount = when (n) { 5 -> 3; 6 -> 4; 7 -> 4; 8 -> 5; 9 -> 5; 10 -> 6; else -> 3 }
        val fascistCount = n - liberalCount - 1
        val roles = MutableList(liberalCount) { SecretHitlerRole.LIBERAL } +
                    MutableList(fascistCount) { SecretHitlerRole.FASCIST } +
                    listOf(SecretHitlerRole.HITLER)
        val shuffled = roles.toMutableList().also { it.shuffle(rng) }
        for ((i, pid) in seatOrder.withIndex()) players[pid]?.role = shuffled[i]
    }

    private fun buildDeck() {
        val deck = MutableList(6) { SecretHitlerPolicy.LIBERAL } +
                   MutableList(11) { SecretHitlerPolicy.FASCIST }
        val shuffled = deck.toMutableList().also { it.shuffle(rng) }
        drawPile.clear(); drawPile.addAll(shuffled)
        discardPile.clear()
    }

    /** Fascists know each other & Hitler. Hitler only knows fascists in 5–6 player games. */
    fun knownAllies(playerId: String): List<String> {
        val me = players[playerId] ?: return emptyList()
        val role = me.role ?: return emptyList()
        val n = seatOrder.size
        return when (role) {
            SecretHitlerRole.LIBERAL -> emptyList()
            SecretHitlerRole.FASCIST -> seatOrder.filter { id ->
                id != playerId && players[id]?.role.let { it == SecretHitlerRole.FASCIST || it == SecretHitlerRole.HITLER }
            }
            SecretHitlerRole.HITLER -> if (n >= 7) emptyList() else seatOrder.filter { id ->
                id != playerId && players[id]?.role == SecretHitlerRole.FASCIST
            }
        }
    }

    fun eligibleChancellorNominees(): List<SecretHitlerPlayer> {
        val aliveCount = alive.size
        return alive.filter { p ->
            p.id != presidentId &&
            p.id != previousChancellorId &&
            !(aliveCount > 5 && p.id == previousPresidentId)
        }
    }

    fun nominateChancellor(targetId: String): Boolean {
        if (phase != SecretHitlerPhase.NOMINATION) return false
        if (eligibleChancellorNominees().none { it.id == targetId }) return false
        chancellorNomineeId = targetId
        electionVotes.clear()
        phase = SecretHitlerPhase.ELECTION
        return true
    }

    fun submitVote(voterId: String, ja: Boolean): Boolean {
        if (phase != SecretHitlerPhase.ELECTION) return false
        val v = players[voterId] ?: return false
        if (!v.alive) return false
        electionVotes[voterId] = ja
        return electionVotes.size >= alive.size
    }

    fun resolveElection(): Boolean {
        if (phase != SecretHitlerPhase.ELECTION) return false
        val yes = electionVotes.values.count { it }
        val no = electionVotes.values.count { !it }
        val passed = yes > no
        lastElectionPassed = passed
        if (passed) {
            previousPresidentId = presidentId
            previousChancellorId = chancellorNomineeId
            chancellorId = chancellorNomineeId
            electionTracker = 0
            val cid = chancellorId
            if (fascistPolicies >= 3 && cid != null && players[cid]?.role == SecretHitlerRole.HITLER) {
                winner = SecretHitlerWinner.FASCIST
                winReason = SecretHitlerWinReason.HITLER_ELECTED_CHANCELLOR
                phase = SecretHitlerPhase.GAME_OVER
                return true
            }
            dealPresidentialHand()
            phase = SecretHitlerPhase.PRESIDENT_DISCARD
            return true
        }
        chancellorNomineeId = null
        chancellorId = null
        electionTracker += 1
        if (electionTracker >= 3) triggerChaos()
        else { advancePresident(); phase = SecretHitlerPhase.NOMINATION }
        return false
    }

    private fun dealPresidentialHand() {
        ensureDeckHasAtLeast(3)
        presidentialHand = drawPile.subList(0, 3).toList()
        repeat(3) { drawPile.removeAt(0) }
    }

    fun presidentDiscard(index: Int): Boolean {
        if (phase != SecretHitlerPhase.PRESIDENT_DISCARD) return false
        if (index !in presidentialHand.indices) return false
        val hand = presidentialHand.toMutableList()
        discardPile.add(hand.removeAt(index))
        chancellorHand = hand
        presidentialHand = emptyList()
        vetoRequested = false
        phase = SecretHitlerPhase.CHANCELLOR_ENACT
        return true
    }

    fun chancellorEnact(index: Int): Boolean {
        if (phase != SecretHitlerPhase.CHANCELLOR_ENACT) return false
        if (index !in chancellorHand.indices) return false
        val hand = chancellorHand.toMutableList()
        val played = hand.removeAt(index)
        discardPile.addAll(hand)
        chancellorHand = emptyList()
        enact(played, byChaos = false)
        return true
    }

    fun chancellorRequestVeto(): Boolean {
        if (phase != SecretHitlerPhase.CHANCELLOR_ENACT || !vetoUnlocked) return false
        vetoRequested = true
        phase = SecretHitlerPhase.VETO_DECISION
        return true
    }

    fun presidentVetoResponse(confirm: Boolean): Boolean {
        if (phase != SecretHitlerPhase.VETO_DECISION) return false
        if (confirm) {
            discardPile.addAll(chancellorHand)
            chancellorHand = emptyList()
            vetoRequested = false
            electionTracker += 1
            if (electionTracker >= 3) triggerChaos()
            else { advancePresident(); phase = SecretHitlerPhase.NOMINATION }
        } else {
            vetoRequested = false
            phase = SecretHitlerPhase.CHANCELLOR_ENACT
        }
        return true
    }

    private fun enact(policy: SecretHitlerPolicy, byChaos: Boolean) {
        lastEnactedPolicy = policy
        lastEnactedByChaos = byChaos
        if (policy == SecretHitlerPolicy.LIBERAL) liberalPolicies += 1 else fascistPolicies += 1
        if (fascistPolicies >= 5) vetoUnlocked = true

        if (liberalPolicies >= 5) {
            winner = SecretHitlerWinner.LIBERAL
            winReason = SecretHitlerWinReason.FIVE_LIBERAL_POLICIES
            phase = SecretHitlerPhase.GAME_OVER
            return
        }
        if (fascistPolicies >= 6) {
            winner = SecretHitlerWinner.FASCIST
            winReason = SecretHitlerWinReason.SIX_FASCIST_POLICIES
            phase = SecretHitlerPhase.GAME_OVER
            return
        }
        if (!byChaos && policy == SecretHitlerPolicy.FASCIST) {
            val p = presidentialPower()
            if (p != null) { enterPower(p); return }
        }
        advancePresident()
        phase = SecretHitlerPhase.NOMINATION
    }

    enum class Power { PEEK, INVESTIGATE, SPECIAL_ELECTION, EXECUTION }

    private fun presidentialPower(): Power? {
        return when (seatOrder.size) {
            5, 6 -> when (fascistPolicies) {
                3 -> Power.PEEK; 4, 5 -> Power.EXECUTION; else -> null
            }
            7, 8 -> when (fascistPolicies) {
                2 -> Power.INVESTIGATE; 3 -> Power.SPECIAL_ELECTION
                4, 5 -> Power.EXECUTION; else -> null
            }
            9, 10 -> when (fascistPolicies) {
                1, 2 -> Power.INVESTIGATE; 3 -> Power.SPECIAL_ELECTION
                4, 5 -> Power.EXECUTION; else -> null
            }
            else -> null
        }
    }

    private fun enterPower(p: Power) {
        when (p) {
            Power.PEEK -> {
                ensureDeckHasAtLeast(3)
                peekedPolicies = drawPile.subList(0, 3).toList()
                phase = SecretHitlerPhase.POLICY_PEEK
            }
            Power.INVESTIGATE -> phase = SecretHitlerPhase.INVESTIGATION
            Power.SPECIAL_ELECTION -> phase = SecretHitlerPhase.SPECIAL_ELECTION
            Power.EXECUTION -> phase = SecretHitlerPhase.EXECUTION
        }
    }

    fun acknowledgePeek(): Boolean {
        if (phase != SecretHitlerPhase.POLICY_PEEK) return false
        peekedPolicies = emptyList()
        advancePresident()
        phase = SecretHitlerPhase.NOMINATION
        return true
    }

    fun investigationTargets(): List<SecretHitlerPlayer> =
        alive.filter { it.id != presidentId && it.id !in investigatedIds }

    fun investigate(targetId: String): Boolean {
        if (phase != SecretHitlerPhase.INVESTIGATION) return false
        val t = players[targetId] ?: return false
        if (!t.alive || t.id == presidentId || t.id in investigatedIds) return false
        pendingInvestigationId = targetId
        lastInvestigation = SecretHitlerInvestigation(targetId, t.role?.party ?: SecretHitlerParty.LIBERAL)
        investigatedIds.add(targetId)
        phase = SecretHitlerPhase.INVESTIGATION_REVEAL
        return true
    }

    fun acknowledgeInvestigation(): Boolean {
        if (phase != SecretHitlerPhase.INVESTIGATION_REVEAL) return false
        pendingInvestigationId = null
        advancePresident()
        phase = SecretHitlerPhase.NOMINATION
        return true
    }

    fun callSpecialElection(targetId: String): Boolean {
        if (phase != SecretHitlerPhase.SPECIAL_ELECTION) return false
        val t = players[targetId] ?: return false
        if (!t.alive || t.id == presidentId) return false
        specialElectionResumeAfter = presidentId
        presidentId = targetId
        chancellorNomineeId = null
        chancellorId = null
        phase = SecretHitlerPhase.NOMINATION
        return true
    }

    fun executionTargets(): List<SecretHitlerPlayer> =
        alive.filter { it.id != presidentId }

    fun executePlayer(targetId: String): Boolean {
        if (phase != SecretHitlerPhase.EXECUTION) return false
        val t = players[targetId] ?: return false
        if (!t.alive || t.id == presidentId) return false
        t.alive = false
        lastExecutedId = targetId
        if (t.role == SecretHitlerRole.HITLER) {
            winner = SecretHitlerWinner.LIBERAL
            winReason = SecretHitlerWinReason.HITLER_EXECUTED
            phase = SecretHitlerPhase.GAME_OVER
            return true
        }
        if (previousChancellorId == targetId) previousChancellorId = null
        if (previousPresidentId == targetId) previousPresidentId = null
        advancePresident()
        phase = SecretHitlerPhase.NOMINATION
        return true
    }

    private fun advancePresident() {
        val resume = specialElectionResumeAfter
        if (resume != null) {
            specialElectionResumeAfter = null
            presidentId = nextAlive(resume)
        } else {
            presidentId?.let { presidentId = nextAlive(it) }
        }
        chancellorNomineeId = null
        chancellorId = null
    }

    private fun nextAlive(from: String): String? {
        val idx = seatOrder.indexOf(from).takeIf { it >= 0 } ?: return alive.firstOrNull()?.id
        val n = seatOrder.size
        for (offset in 1..n) {
            val cand = seatOrder[(idx + offset) % n]
            val p = players[cand]
            if (p != null && p.alive) return cand
        }
        return null
    }

    private fun triggerChaos() {
        ensureDeckHasAtLeast(1)
        val top = drawPile.removeAt(0)
        enact(top, byChaos = true)
        if (phase == SecretHitlerPhase.GAME_OVER) return
        electionTracker = 0
        previousChancellorId = null
        previousPresidentId = null
    }

    private fun ensureDeckHasAtLeast(k: Int) {
        if (drawPile.size >= k) return
        val combined = (drawPile + discardPile).toMutableList()
        combined.shuffle(rng)
        drawPile.clear(); drawPile.addAll(combined)
        discardPile.clear()
    }
}
